sys-utils/setarch.c

```diff
@@ -542,9 +565,33 @@ int main(int argc, char *argv[])
 			} else
 				warnx(_("unrecognized option '--list'"));
 			goto error_getopts;
+		case 'p':
+			if (!archwrapper) {
+				target_pid = strtos32_or_err(optarg, _("invalid PID argument"));
+				if (target_pid <= 0)
+					errx(EXIT_FAILURE, _("out of range value for pid specification: %d"),
+					     target_pid);
+				if (do_show) {
+					show_process_personality(target_pid);
+					return EXIT_SUCCESS;
+				}
+			} else
+				warnx(_("unrecognized option '%s'"), argv[optind - 1]);
+			goto error_getopts;
 		case OPT_SHOW:
 			if (!archwrapper) {
-				if (!optarg || strcmp(optarg, "current") == 0)
+				if (!optarg) {
+					if (target_pid > 0) {
+						/* --pid=PID is given already. */
+						show_process_personality(target_pid);
```

Comment:
You should not rely on another option (`--pid` in this case). You're trying to fix it by `if (do_show)` in the `case 'p``, but it is not an elegant code pattern. It's better to set `do_show=1` only and do everything after the getopt_long() block.